File: frontendiquenobot/src/types/product.ts
import type { ProductStatus } from './enums'

export interface CategoryDto {
  id: string
  name: string
  description: string
  iconUrl: string
  displayOrder: number
  active: boolean
  productCount: number
  createdAt: string
  updatedAt: string
}

export interface CreateCategoryRequest {
  name: string
  description?: string
  iconUrl?: string
  displayOrder?: number
  active?: boolean
}

export interface ProductDto {
  id: string
  sku: string
  name: string
  description: string
  shortDescription: string
  price: number
  compareAtPrice: number
  costPrice: number
  category: CategoryDto | null
  stockQuantity: number
  lowStockThreshold: number
  status: ProductStatus
  imageUrl: string
  images: string
  weight: number | null
  width: number | null
  height: number | null
  length: number | null
  featured: boolean
  tags: string
  available: boolean
  lowStock: boolean
  hasDiscount: boolean
  discountAmount: number
  discountPercentage: number
  createdAt: string
  updatedAt: string
}

export interface CreateProductRequest {
  sku?: string
  name: string
  description?: string
  shortDescription?: string
  price: number
  compareAtPrice?: number
  costPrice?: number
  categoryId?: string
  stockQuantity: number
  lowStockThreshold?: number
  status: ProductStatus
  imageUrl?: string
  images?: string
  weight?: number
  width?: number
  height?: number
  length?: number
  featured?: boolean
  tags?: string
}

File: src/main/java/com/iquenobot/product/domain/entity/Product.java
package com.iquenobot.product.domain.entity;

import com.iquenobot.shared.common.SoftDeletableEntity;
import com.iquenobot.shared.enums.ProductStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.Filter;
import org.hibernate.annotations.ParamDef;

import java.math.BigDecimal;

@Entity
@Table(name = "products")
@Filter(name = "tenantFilter", condition = "tenant_id = :tenantId")
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@AllArgsConstructor
public class Product extends SoftDeletableEntity {

    @Column(name = "sku", length = 100)
    private String sku;

    @Column(name = "name", nullable = false, length = 200)
    private String name;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "short_description", length = 500)
    private String shortDescription;

    @Column(name = "price", nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    @Column(name = "compare_at_price", precision = 10, scale = 2)
    private BigDecimal compareAtPrice;

    @Column(name = "cost_price", precision = 10, scale = 2)
    private BigDecimal costPrice;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @Column(name = "stock_quantity", nullable = false)
    private Integer stockQuantity = 0;

    @Column(name = "low_stock_threshold")
    private Integer lowStockThreshold = 10;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private ProductStatus status;

    @Column(name = "image_url", length = 1000)
    private String imageUrl;

    @Column(name = "images", columnDefinition = "TEXT")
    private String images; // JSON array of image URLs

    @Column(name = "weight")
    private BigDecimal weight;

    @Column(name = "width")
    private BigDecimal width;

    @Column(name = "height")
    private BigDecimal height;

    @Column(name = "length")
    private BigDecimal length;

    @Column(name = "is_featured", nullable = false)
    private boolean featured = false;

    @Column(name = "tags", length = 500)
    private String tags; // JSON array of tags

    @Column(name = "metadata", columnDefinition = "TEXT")
    private String metadata; // JSON for custom fields

    // Business methods
    public boolean isAvailable() {
        return !isDeleted() && status == ProductStatus.ACTIVE && stockQuantity > 0;
    }

    public boolean isLowStock() {
        return lowStockThreshold != null && stockQuantity <= lowStockThreshold;
    }

    public boolean hasDiscount() {
        return compareAtPrice != null && compareAtPrice.compareTo(price) > 0;
    }

    public BigDecimal getDiscountAmount() {
        if (!hasDiscount()) {
            return BigDecimal.ZERO;
        }
        return compareAtPrice.subtract(price);
    }

    public BigDecimal getDiscountPercentage() {
        if (!hasDiscount()) {
            return BigDecimal.ZERO;
        }
        return getDiscountAmount()
                .multiply(BigDecimal.valueOf(100))
                .divide(compareAtPrice, 2, java.math.RoundingMode.HALF_UP);
    }

    public void decreaseStock(int quantity) {
        if (quantity > stockQuantity) {
            throw new IllegalStateException("Stock insuficiente");
        }
        this.stockQuantity -= quantity;
        if (this.stockQuantity == 0) {
            this.status = ProductStatus.OUT_OF_STOCK;
        }
    }

    public void increaseStock(int quantity) {
        this.stockQuantity += quantity;
        if (this.status == ProductStatus.OUT_OF_STOCK && this.stockQuantity > 0) {
            this.status = ProductStatus.ACTIVE;
        }
    }
}